src/redisearch_rs/rqe_iterators/src/optional.rs

```diff
@@ -0,0 +1,193 @@
+            child: Some(child),
+        }
+    }
+}
+
+impl<'index, I> RQEIterator<'index> for Optional<'index, I>
+where
+    I: RQEIterator<'index>,
+{
+    fn read(&mut self) -> Result<Option<&mut RSIndexResult<'index>>, RQEIteratorError> {
+        if self.at_eof() {
+            return Ok(None);
+        }
+
+        let maybe_real = self
+            .child
+            .as_mut()
+            .map(|child| {
+                if child.last_doc_id() != self.result.doc_id {
+                    return Ok(None);
+                }
+                child.read()
+            })
+            .transpose()?
+            .flatten();
+
+        self.result.doc_id += 1;
+
+        if let Some(real) = maybe_real {
+            debug_assert_eq!(
+                self.result.doc_id, real.doc_id,
+                "reads are expected to be always sequential"
+            );
+
+            real.weight = self.weight;
+            return Ok(Some(real));
+        }
+
+        Ok(Some(&mut self.result))
+    }
+
+    // C-Code: SkipTo for OPTIONAL iterator - Non-optimized version.
+    // Skip to a specific docId. If the child has a hit on this docId, return it.
+    // Otherwise, return a virtual hit.
+    fn skip_to(
+        &mut self,
+        doc_id: t_docId,
+    ) -> Result<Option<SkipToOutcome<'_, 'index>>, RQEIteratorError> {
+        debug_assert!(doc_id > self.result.doc_id);
+
+        if doc_id > self.max_doc_id || self.at_eof() {
+            self.result.doc_id = self.max_doc_id;
+            return Ok(None);
+        }
+
+        if let Some(child) = &mut self.child
+            && doc_id > child.last_doc_id()
+            && let Some(SkipToOutcome::Found(real)) = child.skip_to(doc_id)?
+            && real.doc_id == doc_id
+        {
+            // Only a "real" hit where a non-aborted child was able to skip
+            // to the desired doc_id (in contrast to being EOF before)
+            // reaches this branch
+            real.weight = self.weight;
+            self.result.doc_id = real.doc_id;
+            return Ok(Some(SkipToOutcome::Found(real)));
+        }
+
+        self.result.doc_id = doc_id;
+        Ok(Some(SkipToOutcome::Found(&mut self.result)))
+    }
+
+    fn revalidate(&mut self) -> Result<RQEValidateStatus<'_, 'index>, RQEIteratorError> {
+        let Some(ref mut child) = self.child else {
+            return Ok(RQEValidateStatus::Ok);
+        };
+        let last_child_doc_id = child.last_doc_id();
+
+        // Revalidate the child iterator (C-Code: step 1)
+        match child.revalidate()? {
+            // Abort: Handle child validation results (but continue processing)
+            // C-Code: step 2
+            RQEValidateStatus::Aborted => {
+                self.child = None; // Drop it so we become fully virtual until max is reached
+                Ok(if last_child_doc_id != self.result.doc_id {
+                    // virtual
+                    self.result.doc_id += 1;
```

Comment:
Can we test this?
See https://app.codecov.io/gh/RediSearch/RediSearch/pull/7046/blob/src/redisearch_rs/rqe_iterators/src/optional.rs?dropdown=coverage#L145